File: program/src/instruction/reporter.rs
//! Reporter operation instructions

use {
    borsh::BorshSerialize,
    solana_program::{
        instruction::{AccountMeta, Instruction},
        pubkey::Pubkey,
        system_program, sysvar,
    },
};

use crate::{
    error::GenericError,
    id,
    instruction::HapiInstruction,
    state::{
        address::get_address_address,
        case::get_case_address,
        community::get_community_address,
        enums::{Category, CategorySet},
        network::get_network_address,
        reporter::get_reporter_address,
    },
    tools::parse_network_path,
};

/// Creates CreateCase instruction
pub fn create_case(
    // Accounts
    reporter: &Pubkey,
    // Args
    community_name: &str,
    case_id: u64,
    case_name: &str,
    categories: &CategorySet,
) -> Result<Instruction, GenericError> {
    let community_address = get_community_address(&community_name);
    let case_address = get_case_address(&community_address, &case_id.to_le_bytes());
    let reporter_address = get_reporter_address(&community_address, &reporter);

    let accounts = vec![
        AccountMeta::new(*reporter, true),
        AccountMeta::new(community_address, false),
        AccountMeta::new_readonly(reporter_address, false),
        AccountMeta::new(case_address, false),
        AccountMeta::new_readonly(system_program::id(), false),
        AccountMeta::new_readonly(sysvar::rent::id(), false),
    ];

    let instruction = HapiInstruction::CreateCase {
        name: case_name.to_string(),
        categories: *categories,
    };

    Ok(Instruction {
        program_id: id(),
        accounts,
        data: instruction.try_to_vec().unwrap(),
    })
}

/// Creates UpdateCase instruction
pub fn update_case(
    // Accounts
    reporter: &Pubkey,
    // Args
    community_name: &str,
    case_id: u64,
    categories: &CategorySet,
) -> Result<Instruction, GenericError> {
    let community_address = get_community_address(&community_name);
    let case_address = get_case_address(&community_address, &case_id.to_le_bytes());
    let reporter_address = get_reporter_address(&community_address, &reporter);

    let accounts = vec![
        AccountMeta::new(*reporter, true),
        AccountMeta::new_readonly(community_address, false),
        AccountMeta::new_readonly(reporter_address, false),
        AccountMeta::new(case_address, false),
    ];

    let instruction = HapiInstruction::UpdateCase {
        categories: *categories,
    };

    Ok(Instruction {
        program_id: id(),
        accounts,
        data: instruction.try_to_vec().unwrap(),
    })
}

/// Creates ReportAddress instruction
pub fn report_address(
    // Accounts
    reporter: &Pubkey,
    // Args
    network_path: &str,
    address: &Pubkey,
    case_id: u64,
    risk: u8,
    category: Category,
) -> Result<Instruction, GenericError> {
    let (community_name, network_name) = parse_network_path(network_path)?;
    let community_address = get_community_address(&community_name);
    let network_address = get_network_address(&community_address, &network_name);
    let address_address = get_address_address(&network_address, address);
    let reporter_address = get_reporter_address(&community_address, &reporter);
    let case_address = get_case_address(&community_address, &case_id.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*reporter, true),
        AccountMeta::new_readonly(community_address, false),
        AccountMeta::new_readonly(network_address, false),
        AccountMeta::new_readonly(reporter_address, false),
        AccountMeta::new_readonly(case_address, false),
        AccountMeta::new(address_address, false),
        AccountMeta::new_readonly(system_program::id(), false),
        AccountMeta::new_readonly(sysvar::rent::id(), false),
    ];

    let instruction = HapiInstruction::ReportAddress {
        address: *address,
        risk,
        case_id,
        category,
    };

    Ok(Instruction {
        program_id: id(),
        accounts,
        data: instruction.try_to_vec().unwrap(),
    })
}

/// Creates ReportAddress instruction
pub fn update_address(
    // Accounts
    reporter: &Pubkey,
    // Args
    network_path: &str,
    address: &Pubkey,
    case_id: u64,
    risk: u8,
    category: Category,
) -> Result<Instruction, GenericError> {
    let (community_name, network_name) = parse_network_path(network_path)?;
    let community_address = get_community_address(&community_name);
    let network_address = get_network_address(&community_address, &network_name);
    let address_address = get_address_address(&network_address, address);
    let reporter_address = get_reporter_address(&community_address, &reporter);
    let case_address = get_case_address(&network_address, &case_id.to_le_bytes());

    let accounts = vec![
        AccountMeta::new(*reporter, true),
        AccountMeta::new_readonly(community_address, false),
        AccountMeta::new_readonly(reporter_address, false),
        AccountMeta::new_readonly(case_address, false),
        AccountMeta::new_readonly(address_address, false),
    ];

    let instruction = HapiInstruction::UpdateAddress {
        risk,
        case_id,
        category,
    };

    Ok(Instruction {
        program_id: id(),
        accounts,
        data: instruction.try_to_vec().unwrap(),
    })
}

File: js/src/instructions/reporter.ts
import {
  AccountMeta,
  PublicKey,
  TransactionInstruction,
} from "@solana/web3.js";

import { u64 } from "../utils";
import {
  Address,
  Case,
  Network,
  Category,
  Community,
  Reporter,
  CaseStatus,
  Categories,
} from "../state";
import {
  categoriesToBitmask,
  categoryToBinary,
  SYSTEM_RENT_KEYS,
} from "./helpers";
import {
  CreateAddressIx,
  CreateCaseIx,
  UpdateAddressIx,
  UpdateCaseIx,
} from "./instructions";

export async function createCaseInstruction({
  programId,
  payer,
  communityName,
  caseId,
  caseName,
  status,
  categories,
}: {
  programId: PublicKey;
  payer: PublicKey;
  communityName: string;
  caseId: u64;
  caseName: string;
  status: CaseStatus;
  categories: Category[];
}): Promise<TransactionInstruction> {
  if (Buffer.from(caseName).length > 28) {
    throw new Error("Case name length should not be over 28 bytes");
  }

  if (status !== CaseStatus.Open && status !== CaseStatus.Closed) {
    throw new Error(`Unknown case status ${status}`);
  }

  categories.forEach((category) => {
    if (Categories.indexOf(category) < 0) {
      throw new Error(`Unknown category ${category}`);
    }
  });

  const [communityAddress] = await Community.getAddress(
    programId,
    communityName
  );

  const [reporterAddress] = await Reporter.getAddress(
    programId,
    communityAddress,
    payer
  );

  const [caseAddress] = await Case.getAddress(
    programId,
    communityAddress,
    caseId
  );

  const ix = new CreateCaseIx({
    caseId,
    status,
    name: caseName,
    categories: categoriesToBitmask(categories),
  });

  const keys: AccountMeta[] = [
    { pubkey: payer, isSigner: true, isWritable: true },
    { pubkey: communityAddress, isSigner: false, isWritable: true },
    { pubkey: reporterAddress, isSigner: false, isWritable: false },
    { pubkey: caseAddress, isSigner: false, isWritable: true },
    ...SYSTEM_RENT_KEYS,
  ];

  const instruction = new TransactionInstruction({
    keys,
    programId,
    data: ix.encode(),
  });

  return instruction;
}

export async function updateCaseInstruction({
  programId,
  payer,
  communityName,
  caseId,
  status,
  categories,
}: {
  programId: PublicKey;
  payer: PublicKey;
  communityName: string;
  caseId: u64;
  status: CaseStatus;
  categories: Category[];
}): Promise<TransactionInstruction> {
  categories.forEach((category) => {
    if (Categories.indexOf(category) < 0) {
      throw new Error(`Unknown category ${category}`);
    }
  });

  if (status !== CaseStatus.Open && status !== CaseStatus.Closed) {
    throw new Error(`Unknown case status ${status}`);
  }

  const [communityAddress] = await Community.getAddress(
    programId,
    communityName
  );

  const [reporterAddress] = await Reporter.getAddress(
    programId,
    communityAddress,
    payer
  );

  const [caseAddress] = await Case.getAddress(
    programId,
    communityAddress,
    caseId
  );

  const ix = new UpdateCaseIx({
    status,
    categories: categoriesToBitmask(categories),
  });

  const keys: AccountMeta[] = [
    { pubkey: payer, isSigner: true, isWritable: true },
    { pubkey: communityAddress, isSigner: false, isWritable: false },
    { pubkey: reporterAddress, isSigner: false, isWritable: false },
    { pubkey: caseAddress, isSigner: false, isWritable: true },
  ];

  const instruction = new TransactionInstruction({
    keys,
    programId,
    data: ix.encode(),
  });

  return instruction;
}

export async function createAddressInstruction({
  programId,
  payer,
  communityName,
  networkName,
  address,
  caseId,
  risk,
  category,
}: {
  programId: PublicKey;
  payer: PublicKey;
  communityName: string;
  networkName: string;
  address: PublicKey;
  caseId: u64;
  risk: number;
  category: Category;
}): Promise<TransactionInstruction> {
  if (Categories.indexOf(category) < 0) {
    throw new Error(`Unknown category: ${category}`);
  }

  if (risk < 0 || risk > 10) {
    throw new RangeError(`Risk must have a value between 0 and 10`);
  }

  const [communityAddress] = await Community.getAddress(
    programId,
    communityName
  );

  const [reporterAddress] = await Reporter.getAddress(
    programId,
    communityAddress,
    payer
  );

  const [networkAddress] = await Network.getAddress(
    programId,
    communityAddress,
    networkName
  );

  const [caseAddress] = await Case.getAddress(
    programId,
    communityAddress,
    caseId
  );

  const [addressAddress] = await Address.getAddress(
    programId,
    networkAddress,
    address
  );

  const ix = new CreateAddressIx({
    address: address.toBytes(),
    risk,
    caseId,
    category: categoryToBinary(category),
  });

  const keys: AccountMeta[] = [
    { pubkey: payer, isSigner: true, isWritable: true },
    { pubkey: communityAddress, isSigner: false, isWritable: true },
    { pubkey: networkAddress, isSigner: false, isWritable: false },
    { pubkey: reporterAddress, isSigner: false, isWritable: false },
    { pubkey: caseAddress, isSigner: false, isWritable: false },
    { pubkey: addressAddress, isSigner: false, isWritable: true },
    ...SYSTEM_RENT_KEYS,
  ];

  const instruction = new TransactionInstruction({
    keys,
    programId,
    data: ix.encode(),
  });

  return instruction;
}

export async function updateAddressInstruction({
  programId,
  payer,
  communityName,
  networkName,
  address,
  caseId,
  risk,
  category,
}: {
  programId: PublicKey;
  payer: PublicKey;
  communityName: string;
  networkName: string;
  address: PublicKey;
  caseId: u64;
  risk: number;
  category: Category;
}): Promise<TransactionInstruction> {
  if (Categories.indexOf(category) < 0) {
    throw new Error(`Unknown category: ${category}`);
  }

  if (risk < 0 || risk > 10) {
    throw new RangeError(`Risk must have a value between 0 and 10`);
  }

  const [communityAddress] = await Community.getAddress(
    programId,
    communityName
  );

  const [reporterAddress] = await Reporter.getAddress(
    programId,
    communityAddress,
    payer
  );

  const [networkAddress] = await Network.getAddress(
    programId,
    communityAddress,
    networkName
  );

  const [caseAddress] = await Case.getAddress(
    programId,
    communityAddress,
    caseId
  );

  const [addressAddress] = await Address.getAddress(
    programId,
    networkAddress,
    address
  );

  const ix = new UpdateAddressIx({
    risk,
    caseId,
    category: categoryToBinary(category),
  });

  const keys: AccountMeta[] = [
    { pubkey: payer, isSigner: true, isWritable: true },
    { pubkey: communityAddress, isSigner: false, isWritable: false },
    { pubkey: networkAddress, isSigner: false, isWritable: false },
    { pubkey: reporterAddress, isSigner: false, isWritable: false },
    { pubkey: caseAddress, isSigner: false, isWritable: false },
    { pubkey: addressAddress, isSigner: false, isWritable: true },
  ];

  const instruction = new TransactionInstruction({
    keys,
    programId,
    data: ix.encode(),
  });

  return instruction;
}
